### PR TITLE
Strict TypeScript Checking

### PR DESCRIPTION
**Pull request by thegecko**
Note: Relies on #39 to be merged first.

This PR borrows a lot of updates from #36 written by @philipphenkel but takes the strict checking further to include:

- noImplicitThis
- strictBindCallApply
- strictNullChecks
- strictPropertyInitialization
- noImplicitReturns

The only strict check left to implement is `strictFunctionTypes` which needs a fix in [@types/w3c-web-usb](https://www.npmjs.com/package/@types/w3c-web-usb) in order to work.
Specifically the [addEventListener](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/w3c-web-usb/index.d.ts#L115) type needs to be compatible with the base `EventTarget`

cc @larsgk

### CODE REVIEW

**Comment by thegecko** [comment]
@philipphenkel could you check this covers your needs from #36?

**Comment by philipphenkel** [comment]
@thegecko Once you this group on PRs (#36, #39 , #40, #41) is merged to master, I will try to replace my own version. As far as I remember, I have a bunch of additional changes. If those are still valid/needed, I will provide a PR from them.

**Comment by thegecko** [comment]
@philipphenkel thanks. I believe #39 - #41 supersedes the changes in #36, could you let me know if anything is missing?

**Comment by thegecko** [comment]
[@types/w3c-web-usb](https://www.npmjs.com/package/@types/w3c-web-usb) has been [updated](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33025) and `dapjs` can now be built with `strict: true`. Update made to enforce this.

File: tsconfig.json [hunk 1]
```diff
@@ -7,9 +7,8 @@
             "dom",
             "es2015.promise"
         ],
-        "noImplicitAny": true,
-        "noImplicitThis": true,
-        "alwaysStrict": true,
+        "strict": true,
```
Comment (arekzaluski) [inline tsconfig.json]: 👍

File: src/dap/adi.ts [hunk 2]
```diff
@@ -65,21 +65,20 @@ export class ADI implements DAP {
     /**
      * Continually run a function until it returns true
      * @param fn The function to run
-     * @param timer The millisecoinds to wait between each run
+     * @param timer The milliseconds to wait between each run
```
Comment (arekzaluski) [inline src/dap/adi.ts]: 👍

File: src/proxy/cmsis-dap.ts [hunk 3]
```diff
@@ -176,15 +176,15 @@ export class CmsisDAP extends EventEmitter implements Proxy {
...
+     * @returns Promise of number or string
      */
     public dapInfo(request: DAPInfoRequest): Promise<number | string> {
         return this.send(DAPCommand.DAP_INFO, new Uint8Array([request]))
         .then(result => {
             const length = result.getUint8(1);
 
             if (length === 0) {
-                return null;
+                throw new Error("DAP Info Failure");
```
Comment (arekzaluski) [inline src/proxy/cmsis-dap.ts]: That's quite a big change.

File: src/proxy/cmsis-dap.ts [hunk 4]
```diff
@@ -363,10 +363,10 @@ export class CmsisDAP extends EventEmitter implements Proxy {
 
             if (typeof portOrOps === "number") {
                 return result.getUint32(3, true);
-            } else {
-                const length = operations.length * 4;
-                return new Uint32Array(result.buffer.slice(3, 3 + length));
             }
+
+            const length = operations.length * 4;
```
Comment (arekzaluski) [inline src/proxy/cmsis-dap.ts]: 👍

**Pull request merged**